service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/CatalogHandlerUtils.java

```diff
@@ -433,39 +446,174 @@ protected TableMetadata commit(TableOperations ops, UpdateTableRequest request)
+                  MetadataUpdate.SetSnapshotRef setSnapshotRef = findSetSnapshotRefUpdate(request);
+
+                  if (assertRefSnapshotId == null || setSnapshotRef == null) {
+                    // This implies the request was not trying to add a snapshot.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // snapshot-id the client expects the table current_snapshot_id
+                  long expectedCurrentSnapshotId = assertRefSnapshotId.snapshotId();
+                  // table current_snapshot_id.
+                  long currentSnapshotId = base.currentSnapshot().snapshotId();
+                  List<MetadataUpdate> metadataUpdates =
+                      generateUpdatesToRemoveNoopSnapshot(
+                          base,
+                          currentSnapshotId,
+                          expectedCurrentSnapshotId,
+                          setSnapshotRef.name());
+
+                  if (metadataUpdates == null || metadataUpdates.isEmpty()) {
+                    // Nothing can be done as this implies that there were not all
+                    // No-op snapshots (REPLACE) between expectedCurrentSnapshotId and
+                    // currentSnapshotId. hence re-throw the exception caught.
+                    throw new ValidationFailureException(e);
+                  }
+
+                  // Set back the ref we wanted to set, back to the snapshot-id
+                  // the client is expecting the table to be at.
+                  metadataBuilder.setBranchSnapshot(
+                      expectedCurrentSnapshotId, setSnapshotRef.name());
+
+                  // apply the remove snapshots update in the current metadata.
+                  // NOTE: we need to setRef to expectedCurrentSnapshotId first and then apply
+                  // remove, as otherwise the remove will drop the reference.
+                  // NOTE: we can skip removing the now orphan base. Its not a hard requirement.
+                  // just something good to do, and not leave for Remove Orphans.
+                  metadataUpdates.forEach((update -> update.applyTo(metadataBuilder)));
+                  // Ref rolled back update correctly to snapshot to be committed parent now.
+                  newBase = metadataBuilder.build();
+                  // move the lastSequenceNumber back, to apply snapshot properly on the
+                  // current-metadata Seq number are considered increasing monotonically
+                  // snapshot over snapshot, the client generates the manifest list and hence
+                  // the sequence number can't be changed for a snapshot the only possible option
+                  // then is to change the sequenceNumber tracked by metadata.json
+                  Class<?> clazz = newBase.getClass();
+                  try {
+                    Field field = clazz.getDeclaredField("lastSequenceNumber");
+                    field.setAccessible(true);
```

Comment:
Can we avoid reflection workarounds for inaccessible fields?

Reply:
Why not use the `Builder`, transfer snapshots manually, and let the builder figure out the sequence number.

Reply:
What prevents us from adding snapshots in order (we have all the data in memory already)? More compex - yes, but also more correct, IMHO. We're essentially creating a new metadata object here.

Reply:
I think its mainly about if we add Snapshot 1 by 1 to the metadata obj and do the set ref calls accordingly it **_might_** loose the snapshot log which infers things from System.currentTime ([here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1326)) nevertheless now i need to have order of time not just history which will complicate things ([here](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1856)), as if create the whole tree and then toposort.

let me know if your feel strongly about this as IMHO its not worth the complexity to introduce w.r.t a reflection hack (which i know is not that clean ?).

Reply:
Well, my interpretation is this is that Iceberg code is not really reusable for our purpose here :shrug: 

From my POV, dealing with metadata as JSON (or having a Polaris-specific POJO for it) is preferable to tweaking `private` fields in Iceberg code.